js/logic.js

```diff
@@ -0,0 +1,862 @@
+    this.type = opts.type;
+    this.detailPredicate = opts.detailPredicate;
+    this.failOnMismatchedDetails = true;
+    this.not = opts.not;
+    this.timeoutMS = 2000;
+    this.resolved = false;
+
+    logic.defineScope(this, 'LogicMatcher');
+
+    var prevPromise = opts.prevPromise || Promise.resolve();
+
+    if (this.not) {
+      this.promise = prevPromise.then(() => {
+        this.capturedLogs.some((event) => {
+          if ((!this.ns || event.namespace === this.ns) &&
+              event.matches(this.type, this.detailPredicate)) {
+            throw new MismatchError(this, event);
+          }
+        });
+      });
+    } else if (this.type) {
+      this.promise = new Promise((resolve, reject) => {
+        // Once any previous match has been resolved,
+        // subscribe to a following match.
+        prevPromise.then(() => {
+
+          var timeoutId = setTimeout(() => {
+            reject(new Error('LogicMatcherTimeout: ' + this));
+          }, this.timeoutMS);
+
+          // Promise chains have "dead spots" in between resolution
+          // callbacks. For instance:
+          //                 [promise1.then]      [promise2.then]
+          //    other events could be logged --^
+          //
+          // We could miss those events in the middle by just setting
+          // up a new listener for each LogicMatcher. Instead, since
+          // every matcher has a pointer to its prevMatcher, we can
+          // just grab the missing logs from there.
+          var removeMatchListener = (event) => {
+            this.resolved = true;
+            this.capturedLogs = []; // Extra events will go here.
+            setTimeout(function() {
+              emitter.removeListener('*', matchFn);
+            }, 0);
+          };
+
+          var matchFn = (event) => {
+            this.capturedLogs.push(event);
+            if (this.resolved) {
+              return;
+            }
+
+            if (this.ns && event.namespace !== this.ns ||
+                event.type !== this.type) {
+              return false; // did not match
+            }
+            if (event.matches(this.type, this.detailPredicate)) {
+              removeMatchListener(event);
+              this.matchedLogs.push(event);
+              clearTimeout(timeoutId);
+              logic(this, 'match', { ns: this.ns,
+                                     type: this.type,
+                                     event: event });
+              resolve(event);
+              return true;
+            } else {
+              if (this.failOnMismatchedDetails) {
+                removeMatchListener(event);
+                reject(new MismatchError(this, event));
+                return true; // matched
+              } else {
+                // Ignore mismatched events; maybe we'll match later.
+              }
+            }
+            return false; // not done yet, didn't find a match
+          };
+
+          emitter.on('*', matchFn);
```

Comment:
Doing the emitter.on inside a then() doesn't seem right for the base-case where prevPromise is Promise.resolve().  Specifically, you're guaranteeing that we won't start listening for events until some future turn of the event loop.  I would suggest that you synchronously invoke the enclosing function when !opts.prevPromise.